### ISSUE TITLE
Erroneous error about incompatible return value in generic function

### ISSUE THREAD

**Issue by OvervCW** [python/mypy#14813]
**Bug Report**

Consider the following sample code:

```python
from typing import TypeVar

T = TypeVar("T")


def identity(value: T) -> T:
    if isinstance(value, int):
        return value + 0
    else:
        return value


print(identity(2))
print(identity(True))
```

**To Reproduce**

Save this code to a file named `test.py` and run `mypy test.py`.

**Expected Behavior**

I would expect this code to pass with no errors.

**Actual Behavior**

Instead, mypy says this:

```
test.py:8: error: Incompatible return value type (got "int", expected "T")  [return-value]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: `mypy 1.0.1 (compiled: yes)`
- Mypy command-line flags: N/A
- Mypy configuration options from `mypy.ini` (and other config files): N/A
- Python version used: `Python 3.11.1`

**Comment by erictraut** [comment]
Mypy is correct in generating an error here. The expression `value + 0` will always produce an `int`, but the conditional check `isinstance(value, int)` will be true for a subclass of `int`.

Consider the following:
```python
class MyInt(int): ...

print(identity(MyInt(3))) # Produces an int result
```

**Comment by OvervCW** [comment]
@erictraut Fair enough, but I get the same error if I change the check to `type(value) is int`.

**Comment by OvervCW** [comment]
> Yeah, the original example is not a bug.

Could you elaborate?

**Comment by vepain** [comment]
I think the below example is also related to this issue:

```py
class A:
    pass

class B(A):
    pass

def f[T: A](t: T) -> T:
    match t:
        case B():
            print("t is B")
            return t  # Incompatible return value type (got "B", expected "T")
        case A():
            return t
```

* **Python:** `3.13`
* **Mypy:** `1.18.2`

**Comment by ilevkivskyi** [comment]
@vepain Hm, I don't see this error anymore on recent master (also it is not the same as the original report).
